pymc/backends/mcbackend.py

```diff
@@ -0,0 +1,249 @@
+        )
+
+        # Copy at selected indices and append them to the new chain.
+        # This may be slow, but NumPyChain currently don't have a batch-insert or slice API.
+        vnames = [v.name for v in nchain.variables.values()]
+        snames = [s.name for s in nchain.sample_stats.values()]
+        for i in indices:
+            draw = self._chain.get_draws_at(i, var_names=vnames)
+            stats = self._chain.get_stats_at(i, stat_names=snames)
+            nchain.append(draw, stats)
+        return ChainRecordAdapter(nchain, self._point_fn, self._statsbj)
+
+    def point(self, idx: int) -> Dict[str, np.ndarray]:
+        return self._chain.get_draws_at(idx, [v.name for v in self._chain.variables.values()])
+
+
+def make_runmeta_and_point_fn(
+    *,
+    initial_point: Mapping[str, np.ndarray],
+    step: Union[CompoundStep, BlockedStep],
+    model: Model,
+) -> Tuple[mcb.RunMeta, PointFunc]:
+    variables, point_fn = get_variables_and_point_fn(model, initial_point)
+
+    sample_stats = [
+        mcb.Variable("tune", "bool"),
+    ]
+
+    # In PyMC the sampler stats are grouped by the sampler.
+    steps = flatten_steps(step)
+    for s, sm in enumerate(steps):
+        for statname, (dtype, shape) in sm.stats_dtypes_shapes.items():
+            sname = f"sampler_{s}__{statname}"
+            svar = mcb.Variable(
+                name=sname,
+                dtype=np.dtype(dtype).name,
+                shape=shape or [],
+                undefined_ndim=shape is None,
+            )
+            sample_stats.append(svar)
+
+    coordinates = [
+        mcb.Coordinate(dname, mcb.npproto.utils.ndarray_from_numpy(np.array(cvals)))
+        for dname, cvals in model.coords.items()
+        if cvals is not None
+    ]
+    meta = mcb.RunMeta(
+        rid=hagelkorn.random(),
+        variables=variables,
+        coordinates=coordinates,
+        sample_stats=sample_stats,
+        data=find_data(model),
+    )
+    return meta, point_fn
+
+
+def init_chain_adapters(
+    *,
+    backend: mcb.Backend,
+    chains: int,
+    initial_point: Mapping[str, np.ndarray],
+    step: Union[CompoundStep, BlockedStep],
+    model: Model,
+) -> List[ChainRecordAdapter]:
+    """Create an McBackend metadata description for the MCMC run.
+
+    Parameters
+    ----------
+    backend
+        An McBackend `Backend` instance.
+    chains
+        Number of chains to initialize.
+    initial_point
+        Dictionary mapping value variable names to initial values.
+    step : CompoundStep or BlockedStep
+        The step method that iterates the MCMC.
+    model : pm.Model
+        The current PyMC model.
+
+    Returns
+    -------
+    adapters
+        Chain recording adapters that wrap McBackend Chains in the PyMC IBaseTrace interface.
+    """
+    meta, point_fn = make_runmeta_and_point_fn(initial_point=initial_point, step=step, model=model)
+    run = backend.init_run(meta)
```

Comment:
For most consistent API we should ideally return the `run` object to the user (`pm.sample(return_mcb=True)`)

Reply:
I will do more refactoring of the `pm.sample` (specifically the part where it converts to `InferenceData` and returns the result) to make it more versatile and less 🍝